innkeeper/client.go

```diff
@@ -0,0 +1,417 @@
+	Id        int64    `json:"id"`
+	CreatedAt string   `json:"createdAt"`
+	DeletedAt string   `json:"deletedAt"`
+	Route     routeDef `json:"route"`
+}
+
+type apiError struct {
+	Message   string `json:"message"`
+	ErrorType string `json:"type"`
+}
+
+type routeDoc string
+
+type Client struct {
+	baseUrl      *url.URL
+	auth         Authentication
+	authToken    string
+	lastModified string
+	httpClient   *http.Client
+	dataChan     chan skipper.RawData
+
+	// store the routes for comparison during
+	// the subsequent polls
+	doc    map[int64]string
+	closer chan interface{}
+	closed chan interface{}
+}
+
+type Options struct {
+	Address        string
+	Insecure       bool
+	PollTimeout    time.Duration
+	Authentication Authentication
+}
+
+// Creates an innkeeper client.
+func Make(o Options) (*Client, error) {
+	u, err := url.Parse(o.Address)
+	if err != nil {
+		return nil, err
+	}
+
+	c := &Client{
+		u, o.Authentication, "", "",
+		&http.Client{Transport: &http.Transport{
+			TLSClientConfig: &tls.Config{InsecureSkipVerify: o.Insecure}}},
+		make(chan skipper.RawData),
+		make(map[int64]string),
+		make(chan interface{}),
+		make(chan interface{})}
+
+	// start a polling loop
+	go func() {
+		to := time.Duration(0)
+		for {
+			select {
+			case <-time.After(to):
+				to = o.PollTimeout
+				if c.poll() {
+					c.dataChan <- toDocument(c.doc)
+				}
+			case <-c.closer:
+				return
+			}
+		}
+	}()
+
+	return c, nil
+}
+
+func (c *Client) createUrl(path ...string) string {
+	u := *c.baseUrl
+	u.Path = strings.Join(append([]string{""}, path...), "/")
+	return (&u).String()
+}
+
+func parseApiError(r io.Reader) (string, error) {
+	message, err := ioutil.ReadAll(r)
+
+	if err != nil {
+		return "", err
+	}
+
+	ae := apiError{}
+	if err := json.Unmarshal(message, &ae); err != nil {
+		return "", err
+	}
+
+	return ae.ErrorType, nil
+}
+
+func isApiAuthError(error string) bool {
+	switch authErrorType(error) {
+	case authErrorPermission, authErrorAuthentication:
+		return true
+	default:
+		return false
+	}
+}
+
+func (c *Client) authenticate() error {
+	t, err := c.auth.Token()
+	if err != nil {
+		return err
+	}
+
+	c.authToken = t
+	return nil
+}
+
+func (c *Client) requestData(authRetry bool, url string) ([]*routeData, error) {
+	req, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	req.Header.Add("Authorization", c.authToken)
+	response, err := c.httpClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
+
+	defer response.Body.Close()
+
+	if response.StatusCode == 401 {
+		apiError, err := parseApiError(response.Body)
+		if err != nil {
+			return nil, err
+		}
+
+		if !isApiAuthError(apiError) {
+			return nil, fmt.Errorf("unknown authentication error: %s", apiError)
+		}
+
+		if !authRetry {
+			return nil, errors.New("authentication failed")
+		}
+
+		err = c.authenticate()
+		if err != nil {
+			return nil, err
+		}
+
+		return c.requestData(false, url)
+	}
+
+	if response.StatusCode >= 400 {
```

Comment:
Hey Luis, what's your recommended approach? What we just quickly discussed here, is that comparing `>= http.StatusBadRequest` is fine, we can distinguish between 40x and 50x, but in general, it is fine. I'll can replace the num literal with the constant form, of course. How should this look in your opinion?